compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -2162,14 +2162,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         } else if !expr_t.is_primitive_ty() {
             self.ban_nonexisting_field(field, base, expr, expr_t);
         } else {
-            type_error_struct!(
+            let field_name = field.to_string();
+            let mut err = type_error_struct!(
                 self.tcx().sess,
                 field.span,
                 expr_t,
                 E0610,
                 "`{expr_t}` is a primitive type and therefore doesn't have fields",
-            )
-            .emit();
+            );
+            if expr_t.is_integral()
```

Comment:
Is it possible for you to check that  the type we're dereferencing with `.` is a numerical literal? Specifically, I don't want it to fix:

```
let x = 1i32;
x.e10;
```

Reply:
Or if it already accounts for this, can you add a unit test for this case?

Reply:
To elaborate on this, I would probably just `span_to_snippet` and check that it parses as a valid number. You should also add a check for 

```
macro_rules! num { () => { 1 } }

fn main() {
    num!().e10;
}
```